presto/dbapi.py

```diff
@@ -236,10 +245,111 @@ def setinputsizes(self, sizes):
     def setoutputsize(self, size, column):
         raise presto.exceptions.NotSupportedError
 
-    # TODO support params
-    def execute(self, operation):
-        self._query = presto.client.PrestoQuery(self._request, sql=operation)
-        result = self._query.execute()
+    def _execute_prepare_statement(self, operation, statement_name):
+        """
+        Prepends the given `operation` with "PREPARE <statement_name> FROM" and
+        executes as a prepare statement.
+
+        :param operation: sql to be executed.
+        :param statement_name: name that will be assigned to the prepare
+            statement.
+
+        :raises presto.exceptions.FailedToObtainAddedPrepareHeader: Error raised
+            when unable to find the 'X-Presto-Added-Prepare' for the PREPARE
+            statement request.
+
+        :return: string representing the value of the 'X-Presto-Added-Prepare'
+            header.
+        """
+        sql = 'PREPARE {statement_name} FROM {operation}'.format(
+            statement_name=statement_name,
+            operation=operation
+        )
+
+        # Send prepare statement. Copy the _request object to avoid poluting the
+        # one that is going to be used to execute the actual operation.
+        query = presto.client.PrestoQuery(copy.deepcopy(self._request), sql=sql)
+        result = query.execute()
+
+        # Iterate until the 'X-Presto-Added-Prepare' header is found or
+        # until there are no more results
+        for _ in result:
+            response_headers = result.response_headers
+
+            if constants.HEADER_ADDED_PREPARE in response_headers:
+                return response_headers[constants.HEADER_ADDED_PREPARE]
+
+        raise presto.exceptions.FailedToObtainAddedPrepareHeader
+
+    def _get_added_prepare_statement_presto_query(
+        self,
+        added_prepare_header,
+        statement_name,
+        params
+    ):
+        sql = 'EXECUTE ' + statement_name + ' USING ' + ','.join(map(str, params))
```

Comment:
This method implicitly plays with the string representation of `params` which can cause issues (large numbers which python prints using scientific notation, arrays etc.).

Reply:
Good point. What do you suggest in this case?

I think it's fair to assume that any complex object in the `params` the  must implement the `__str__` function  in order to be treated properly. Otherwise, they should be parsed/converted on the user's implementation before appending them to the `params`.

Reply:
@harph You are correct. I think we can leave as is in that case but clarify this in the docstring of the method to avoid people like me being caught by surprise.

**That still leaves one concern and that is regarding `params` whose values themselves include a `,`.**

Been a little long since I used Python with any DB. Almost every DB interprets the values passes as parameters as `LITERALS` by converting to a string.

Reply:
https://github.com/prestosql/presto-python-client/pull/16/files#diff-e6510cf8abf3e99495eee2b0d39cd839R54 might be useful.

@harph

Reply:
>I think it's fair to assume that any complex object in the params the must implement the __str__ function in order to be treated properly.

__str__ has different contract and different requirements than Presto's parser.

I'd rather err on the side of not supporting a type, rather than call str on something I do not know.

I assume that supporting totally custom types is not a requirement, so we do not have to worry about them.